consensus/state_processing/src/per_epoch_processing/altair/participation_cache.rs

```diff
@@ -0,0 +1,409 @@
+    }
+}
+
+/// Caches the participation values for one epoch (either the previous or current).
+#[derive(PartialEq, Debug)]
+struct SingleEpochParticipationCache {
+    /// Maps an active validator index to their participation flags.
+    ///
+    /// To reiterate, only active and unslashed validator indices are stored in this map.
+    ///
+    /// ## Note
+    ///
+    /// It would be ideal to maintain a reference to the `BeaconState` here rather than copying the
+    /// `ParticipationFlags`, however that would cause us to run into mutable reference limitations
+    /// upstream.
+    unslashed_participating_indices: HashMap<usize, ParticipationFlags>,
+    /// Stores the sum of the balances for all validators in `self.unslashed_participating_indices`
+    /// for all flags in `NUM_FLAG_INDICES`.
+    ///
+    /// A flag balance is only incremented if a validator is that flag set.
+    total_flag_balances: [Balance; NUM_FLAG_INDICES],
+    /// Stores the sum of all balances of all validators in `self.unslashed_participating_indices`
+    /// (regardless of which flags are set).
+    total_active_balance: Balance,
+}
+
+impl SingleEpochParticipationCache {
+    fn new(hashmap_len: usize, spec: &ChainSpec) -> Self {
+        let zero_balance = Balance::zero(spec.effective_balance_increment);
+
+        Self {
+            unslashed_participating_indices: HashMap::with_capacity(hashmap_len),
+            total_flag_balances: [zero_balance; NUM_FLAG_INDICES],
+            total_active_balance: zero_balance,
+        }
+    }
+
+    /// Shrink any allocations to the smallest necessary size.
+    fn shrink_to_fit(&mut self) {
+        self.unslashed_participating_indices.shrink_to_fit()
+    }
+
+    /// Returns the total balance of attesters who have `flag_index` set.
+    fn total_flag_balance(&self, flag_index: usize) -> Result<u64, Error> {
+        self.total_flag_balances
+            .get(flag_index)
+            .map(Balance::get)
+            .ok_or(Error::InvalidFlagIndex(flag_index))
+    }
+
+    /// Returns `true` if `val_index` is active, unslashed and has `flag_index` set.
+    ///
+    /// ## Errors
+    ///
+    /// May return an error if `flag_index` is out-of-bounds.
+    fn has_flag(&self, val_index: usize, flag_index: usize) -> Result<bool, Error> {
+        if let Some(participation_flags) = self.unslashed_participating_indices.get(&val_index) {
+            participation_flags
+                .has_flag(flag_index)
+                .map_err(|_| Error::InvalidFlagIndex(flag_index))
+        } else {
+            Ok(false)
+        }
+    }
+
+    /// Process an **active** validator, reading from the `state` with respect to the
+    /// `relative_epoch`.
+    ///
+    /// ## Errors
+    ///
+    /// - The provided `state` **must** be Altair. An error will be returned otherwise.
+    ///
+    /// ## Warning
+    ///
+    /// - It is a logic error to provide an inactive validator to this function.
```

Comment:
Seeing as it's a cheap check and we're already fetching the `Validator` struct, I think we may as well defensively check that the validator is actually active (and return an error if they're not).